src/lib/semaphore.js

```diff
@@ -0,0 +1,29 @@
+export default class Semaphore {
+  constructor(maxResources) {
+    this.maxResources = maxResources;
```

Comment:
would love to see some comments as I had to stare at this for awhile to understand it. Also not sure if "maxResources" makes sense as a name or if it should be more generic

Reply:
OK will add comments and rename to be more generic